### PR TITLE
Multiswaps

### PR DESCRIPTION
**Pull request by iStrike7**

### CODE REVIEW

File: pallets/xyk/src/lib.rs [hunk 1]
```diff
@@ -2067,6 +2217,141 @@ impl<T: Config> PreValidateSwaps for Pallet<T> {
...
+	> {
+		// Ensure not buying zero amount
+		ensure!(!final_bought_asset_amount.is_zero(), Error::<T>::ZeroAmount,);
+		ensure!(!max_amount_in.is_zero(), Error::<T>::ZeroAmount,);
+
+		ensure!(swap_token_list.len() > 2_usize, Error::<T>::MultiswapShouldBeAtleastTwoHops);
+		// Unwraps are fine due to above ensure
+		let sold_asset_id = *swap_token_list.get(0).unwrap();
+		let bought_asset_id = *swap_token_list.get(1).unwrap();
+
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:2247]:
```
		ensure!(swap_token_list.len() > 2_usize, Error::<T>::MultiswapShouldBeAtleastTwoHops);
		// Unwraps are fine due to above ensure
		let sold_asset_id = *swap_token_list.get(0).unwrap();
		let bought_asset_id = *swap_token_list.get(1).unwrap();
    ```
    i would rather use below to completly get rid of unwraps

```
		let sold_asset_id = *swap_token_list.get(0).ok_or(Error::<T>::MultiswapShouldBeAtleastTwoHops)?
		let bought_asset_id = *swap_token_list.get(1).ok_or(Error::<T>::MultiswapShouldBeAtleastTwoHops)?;
```

File: pallets/xyk/src/lib.rs [hunk 2]
```diff
@@ -2369,10 +2654,193 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
...
+			swap_token_list.clone(),
+			sold_asset_amount,
+			min_amount_out,
+		)?;
+
+		// First execute all atomic swaps in a storage layer
+		// And then the finally bought amount is compared
+		// The bool in error represents if the fail is due to bad final slippage
+		let res_multiswap =
+			frame_support::storage::with_storage_layer(|| -> Result<Balance, DispatchError> {
```
Comment (majusko) [inline pallets/xyk/src/lib.rs:2691]: code style comment:

Let's move the whole functionality of `let res_multiswap =` calculation to a separate method.

File: pallets/xyk/src/lib.rs [hunk 3]
```diff
@@ -2533,7 +3001,204 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
...
+			&sender,
+			swap_token_list.clone(),
+			bought_asset_amount,
+			max_amount_in,
+		)?;
+
+		// First execute all atomic swaps in a storage layer
+		// And then the finally sold amount is compared
+		// The bool in error represents if the fail is due to bad final slippage
+		let res_multiswap =
```
Comment (majusko) [inline pallets/xyk/src/lib.rs:3033]: code style comment:

Let's move the whole functionality of let res_multiswap = calculation to a separate method.

**Review by majusko — CHANGES_REQUESTED** [review]

File: pallets/xyk/src/lib.rs [hunk 4]
```diff
@@ -2067,6 +2217,141 @@ impl<T: Config> PreValidateSwaps for Pallet<T> {
...
+				} else if x > y {
+					Some((x, y))
+				} else
+				// x < y
+				{
+					Some((y, x))
+				}
+			})
+			.collect();
+
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:2278]: if you collect directly to hashset, you can avoide few unecescary allocations, and separate dedup call

File: pallets/xyk/src/lib.rs [hunk 5]
```diff
@@ -2067,6 +2217,141 @@ impl<T: Config> PreValidateSwaps for Pallet<T> {
...
+		// Cannot use multiswap twice on the same pool
+		let mut pair_sold_assets = swap_token_list.clone();
+		let _ = pair_sold_assets.pop();
+
+		let mut pair_bought_assets = swap_token_list.clone();
+		let _ = pair_bought_assets.remove(0);
+
+		let atomic_pairs: Vec<(TokenId, TokenId)> =
+			pair_sold_assets.into_iter().zip(pair_bought_assets.into_iter()).collect();
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:2262]: remove(0) on vec is not very efficient, consider using 
```
		let atomic_pairs: Vec<(TokenId, TokenId)> =
			swap_token_list.iter().skip(1).zip(swap_token_list.iter()).collect();

```
instead

File: pallets/xyk/src/lib.rs [hunk 6]
```diff
@@ -2369,10 +2654,193 @@ impl<T: Config> XykFunctionsTrait<T::AccountId> for Pallet<T> {
...
+
+				// pre_validate has already confirmed that swap_token_list.len()>1
+				let mut pair_sold_assets = swap_token_list.clone();
+				let _ = pair_sold_assets.pop();
+
+				let mut pair_bought_assets = swap_token_list.clone();
+				let _ = pair_bought_assets.remove(0);
+
+				let atomic_pairs: Vec<(TokenId, TokenId)> =
+					pair_sold_assets.into_iter().zip(pair_bought_assets.into_iter()).collect();
```
Comment (mateuszaaa) [inline pallets/xyk/src/lib.rs:2711]: remove(0) on vec is not very efficient, consider using 
```
		let atomic_pairs: Vec<(TokenId, TokenId)> =
			swap_token_list.iter().skip(1).zip(swap_token_list.iter()).collect();

```
instead

**Review by mateuszaaa — CHANGES_REQUESTED** [review]

File: pallets/xyk/src/lib.rs [hunk 7]
```diff
@@ -675,6 +688,26 @@ pub mod pallet {
 			Ok(().into())
 		}
 
+		#[pallet::weight(<<T as Config>::WeightInfo>::multiswap_sell_asset(swap_token_list.len() as u32))]
```
Comment (majusko) [inline pallets/xyk/src/lib.rs:691]: Missing docs:

Please add the docs to all `pub fn` for our docs autogeneration, the same as it is done in the Bootstrap pallet.

**Review by majusko — CHANGES_REQUESTED** [review]

**Review by mateuszaaa — CHANGES_REQUESTED** [review]
please assign full-benchmarks label so ci can output new benchmarks for newly added functions

**Review by majusko — APPROVED** [review]
🙌 nice work!

**Review by mateuszaaa — APPROVED** [review]
lgmt, lets wait for ci and merge it

**Review by mateuszaaa — APPROVED** [review]
lgtm

**Pull request merged**